### PR TITLE
Cards management module

### LINKED ISSUE #8: Management Module - Normal Cards
Module is redesigned, and following functions are merged.

int initialize_game(void);
void initialize_players(void);
int initialize_cards(void);
int deal_cards(void);
int add_card_at_beginning(struct DECK** head, struct CARD card);
int add_card_at_end(struct DECK* head, struct CARD card);
struct DECK* remove_first_card_at_beginning(struct DECK** head);
struct DECK* find_playable_card(enum PlayerType player);
void display_cards_list(struct DECK *list_ptr);
int get_pile_length(struct DECK* deck_ptr);
bool   is_pile_empty(enum CardPile pile);
bool is_playable_card(struct CARD card);
int sort_cards_on_hand(enum PlayerType sort_player);
void swap_cards(struct CARD* a, struct CARD* b);
struct DECK *remove_first_playable_card(struct DECK** head);
int discard_card(enum PlayerType player, int* ptr_post_condition);
int draw_cards(int num_draw_cards, enum PlayerType player);
struct CARD draw_one_card(enum PlayerType player);
enum PlayerType get_game_winner(void);

### PR DESCRIPTION
**Pull request by MiaYWei**
Cards management module:

void initialize_players(void);
int initialize_cards(void);
int deal_cards(void);
int add_card_at_beginning(struct DECK** head, struct CARD card);
int add_card_at_end(struct DECK* head, struct CARD card);
struct DECK* remove_first_card_at_beginning(struct DECK** head);
struct DECK* find_playable_card(enum PlayerType player);
void display_cards_list(struct DECK *list_ptr);
int get_pile_length(struct DECK* deck_ptr);
bool is_pile_empty(enum CardPile pile);
bool is_playable_card(struct CARD card);
int sort_cards_on_hand(enum PlayerType sort_player);
void swap_cards(struct CARD* a, struct CARD* b);
struct DECK *remove_first_playable_card(struct DECK** head);
int discard_card(enum PlayerType player, int* ptr_post_condition);
int draw_cards(int num_draw_cards, enum PlayerType player);
struct CARD draw_one_card(enum PlayerType player);
enum PlayerType get_game_winner(void);

### CODE REVIEW

**Review by aditibhardwaj98 — APPROVED** [review]

File: src/cards_management.h [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+/**
+ * @brief Creates a new node and adds it at the end of the linked list.
+ *
+ * @param[in/out] head: pointer which points to the head of the linked list.
+ * @param[in] data: int type variable which the data of the new node
+ */
+int add_card_at_end(struct DECK* head, struct CARD card);
+
+/**
+ * @brief remove the first card st the beginning of the card list
```
Comment (krithigasr) [inline src/cards_management.h]: Typo - first card 'at' the beginning

File: src/cards_management.h [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+* Release 1: 10 cards per colour;
+* Release 2: 4 Skip cards; 4 Draw One cards; 2 Wild cards; 2 Wild Draw Two cards*/
+#define MAX_CARDS_NUM   40
+#define LINE_SIZE       1000         /*LINESIZE to read input */
+#define MALLOC_FAIL     1
+#define SUCCESS         0
+#define PLAYERS_NUM     2
+#define DEAL_CARDS_NUM  5
+
+enum PlayerType { HUMAN, COMPUTER};
```
Comment (krithigasr) [inline src/cards_management.h]: Like discussed, can you please add 'tydef' to struct and enum.

File: src/cards_management.h [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+enum Direction { CLOCKWISE, COUNTER_CLOCKWISE };
```
Comment (krithigasr) [inline src/cards_management.h]: What are we using Direction for?

File: src/cards_management.h [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+ * @return const Deck*  pointer type variable, which points to the playable card.
+ */
+struct DECK* find_playable_card(enum PlayerType player);
+
+/**
+ * @brief Displays the detailed card infomation from the card list 
+ * 
+ * @param list_ptr The pointer which points to the beginning of the cards list
+ */
+void display_cards_list(struct DECK *list_ptr);
```
Comment (krithigasr) [inline src/cards_management.h]: Is this used to initially display the set of cards to each player?

**Review by krithigasr — COMMENTED** [review]
Like discussed, can you please add 'tydef' to struct and enum.

**Review by krithigasr — APPROVED** [review]

**Review by HongBo713 — APPROVED** [review]

**Pull request merged**